cmd/endpoint.go

```diff
@@ -0,0 +1,430 @@
+		if i == 0 {
+			endpointType = endpoint.Type()
+			scheme = endpoint.Scheme
+		} else if endpoint.Type() != endpointType {
+			return nil, fmt.Errorf("mixed style endpoints are not supported")
+		} else if endpoint.Scheme != scheme {
+			return nil, fmt.Errorf("mixed scheme is not supported")
+		}
+
+		arg = endpoint.String()
+		if uniqueArgs.Contains(arg) {
+			return nil, fmt.Errorf("duplicate endpoints found")
+		}
+		uniqueArgs.Add(arg)
+
+		endpoints = append(endpoints, endpoint)
+	}
+
+	sort.Sort(endpoints)
+
+	return endpoints, nil
+}
+
+// CreateEndpoints - validates and creates new endpoints for given args.
+func CreateEndpoints(serverAddr string, args ...string) (string, EndpointList, SetupType, error) {
+	var endpoints EndpointList
+	var setupType SetupType
+	var err error
+
+	// Check whether serverAddr is valid for this host.
+	isServerAddrEmpty := (serverAddr == "")
+	if !isServerAddrEmpty {
+		if err = CheckLocalServerAddr(serverAddr); err != nil {
+			return serverAddr, endpoints, setupType, err
+		}
+	}
+
+	// Normalize server address.
+	serverAddrHost, serverAddrPort := mustSplitHostPort(serverAddr)
+	serverAddr = net.JoinHostPort(serverAddrHost, serverAddrPort)
+
+	// For single arg, return FS setup.
+	if len(args) == 1 {
+		var endpoint Endpoint
+		if endpoint, err = NewEndpoint(args[0]); err == nil {
+			if endpoint.Type() == PathEndpointType {
+				endpoints = append(endpoints, endpoint)
+				setupType = FSSetupType
+			} else {
+				err = fmt.Errorf("use path style endpoint for FS setup")
+			}
+		}
+
+		return serverAddr, endpoints, setupType, err
+	}
+
+	// Convert args to endpoints
+	if endpoints, err = NewEndpointList(args...); err != nil {
+		return serverAddr, endpoints, setupType, err
+	}
+
+	// Return XL setup when all endpoints are path style.
+	if endpoints[0].Type() == PathEndpointType {
+		setupType = XLSetupType
+		return serverAddr, endpoints, setupType, nil
+	}
+
+	// Here all endpoints are URL style.
+
+	// Error out if same path is exported by different ports on same server.
+	{
+		hostPathMap := make(map[string]set.StringSet)
+		for _, endpoint := range endpoints {
+			hostPort := endpoint.Host
+			path := endpoint.Path
+			host, _ := mustSplitHostPort(hostPort)
+
+			pathSet, ok := hostPathMap[host]
+			if !ok {
+				pathSet = set.NewStringSet()
+			}
+
+			if pathSet.Contains(path) {
+				return serverAddr, endpoints, setupType, fmt.Errorf("same path can not be served from different port")
+			}
+
+			pathSet.Add(path)
+			hostPathMap[host] = pathSet
+		}
+	}
+
+	// Normalized args is used below if URL style endpoint is used for XL.
+	normArgs := make([]string, len(args))
+
+	// Get unique hosts.
+	var uniqueHosts []string
+	{
+		sset := set.NewStringSet()
+		for i, endpoint := range endpoints {
+			sset.Add(endpoint.Host)
+			normArgs[i] = endpoint.Path
+		}
+		uniqueHosts = sset.ToSlice()
+	}
+
+	// URL style endpoints are used for XL.
+	if len(uniqueHosts) == 1 {
```

Comment:
this code is not needed, check https://github.com/balamurugana/minio/pull/5 7b7f24b for details.